### PR TITLE
Minimum CMake downgrade for Linux

### LINKED ISSUE #32199: Requiring CMake versions that aren't in the default package repository
According to https://github.com/dotnet/runtime/blob/master/docs/workflow/requirements/linux-requirements.md#toolchain-setup

> Building the repo requires CMake 3.14.2 or newer on Linux. Add Kitware's APT feed to your configuration for a newer version of CMake.

This is going to cause problems if/when we want to ship .NET 5 in Ubuntu 18's default package feed. In order to submit our product to these feeds, we can't rely on versions of tools that are in external package feeds.

We should allow our product to be built by toolchains that are available by default in major Linux distros, and not require external feeds in order to build.

/cc @dseefeld @dleeapho

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Downgrade the minimum version of CMake to 3.6.2 for our Linux builds and change our CMake scripts appropriately:

Currently completed steps:

- Provide our own implementation of `add_compile_definitions` since it is easily polyfillable
- Move away from `add_link_options` since it is not easy to create our own implementation of that functions the same. Move back to setting linker options via the flag properties.
- Manually enable PIE/PIC on Linux/Mac (this support was added in CMake 3.9)
- Manually check for IPO support on Linux/Mac (this support was added in CMake 3.13)
- Revert to the old CMake command line (this support was added in CMake 3.10)

So far I've only undone the changes that I made in dotnet/coreclr#26980 that I know are required. @janvorli can you help out with some of the other issues that we'll encounter when lowering the CMake version? I don't know all of the corner case behaviors that we've started to depend on in the last 10ish months.

Validation that we actually build on CMake 3.6.2 will come when we get a new Docker image has CMake 3.6.2 (and no newer) installed (@jashook is working on this).

Contributes to #38755

Depends on dotnet/arcade#5770

cc: @dleeapho @janvorli @jashook @omajid @jkotas @tmds @wfurt @RheaAyase

### CODE REVIEW

File: eng/native/configurecompiler.cmake [hunk 1]
```diff
@@ -157,11 +161,12 @@ elseif (CLR_CMAKE_HOST_UNIX)
...
       # -O1: optimization level used instead of -O0 to avoid compile error "invalid operand for inline asm constraint"
-      add_compile_definitions("$<$<OR:$<CONFIG:DEBUG>,$<CONFIG:CHECKED>>:${CLR_SANITIZE_CXX_OPTIONS};-fdata-sections;--ffunction-sections;-O1>")
-      add_link_options($<$<AND:$<OR:$<CONFIG:DEBUG>,$<CONFIG:CHECKED>>,$<STREQUAL:$<TARGET_PROPERTY:TYPE>,EXECUTABLE>>:${CLR_SANITIZE_LINK_OPTIONS}>)
-
+      add_compile_options("$<$<OR:$<CONFIG:DEBUG>,$<CONFIG:CHECKED>>:${CLR_SANITIZE_CXX_OPTIONS};-fdata-sections;--ffunction-sections;-O1>")
+      add_linker_flag("${CLR_SANITIZE_LINK_OPTIONS}" DEBUG)
+      add_linker_flag("${CLR_SANITIZE_LINK_OPTIONS}" CHECKED)
       # -Wl and --gc-sections: drop unused sections\functions (similar to Windows /Gy function-level-linking)
-      add_link_options("$<$<AND:$<OR:$<CONFIG:DEBUG>,$<CONFIG:CHECKED>>,$<STREQUAL:$<TARGET_PROPERTY:TYPE>,SHARED_LIBRARY>>:${CLR_SANITIZE_LINK_OPTIONS};-Wl,--gc-sections>")
+      add_linker_flag("-Wl,--gc-sections" DEBUG)
```
Comment (janvorli) [inline eng/native/configurecompiler.cmake:168]: A nit - maybe the add_linker_flag could be made to accept multiple configurations so that we can write
```cmake
add_linker_flag("-Wl,--gc-sections" DEBUG CHECKED)
```

File: eng/native/configurecompiler.cmake [hunk 2]
```diff
@@ -173,15 +178,22 @@ endif(MSVC)
 #       ./build-native.sh cmakeargs "-DCLR_ADDITIONAL_COMPILER_OPTIONS=<...>" cmakeargs "-DCLR_ADDITIONAL_LINKER_FLAGS=<...>"
 #
 if(CLR_CMAKE_HOST_UNIX)
-    add_link_options(${CLR_ADDITIONAL_LINKER_FLAGS})
+  foreach(ADDTL_LINKER_FLAG ${CLR_ADDITIONAL_LINKER_FLAGS})
+  add_linker_flag(${ADDTL_LINKER_FLAG})
```
Comment (janvorli) [inline eng/native/configurecompiler.cmake:182]: A nit - can you please indent the body?

File: eng/native/configurecompiler.cmake [hunk 3]
```diff
@@ -173,15 +178,22 @@ endif(MSVC)
...
 if(CLR_CMAKE_HOST_LINUX)
   add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
-  add_link_options(-Wl,--build-id=sha1 -Wl,-z,relro,-z,now)
+  add_linker_flag(-Wl,--build-id=sha1)
+  add_linker_flag(-Wl,-z,relro,-z,now)
 elseif(CLR_CMAKE_HOST_FREEBSD)
   add_compile_options($<$<COMPILE_LANGUAGE:ASM>:-Wa,--noexecstack>)
-  add_link_options(LINKER:--build-id=sha1)
+  if (CMAKE_CXX_COMPILER_ID STREQUAL Clang)
+    add_linker_flag("-Xlinker --build-id=sha1")
```
Comment (janvorli) [inline eng/native/configurecompiler.cmake:193]: Clang should accept the -Wl syntax too

File: src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt [hunk 4]
```diff
@@ -33,7 +33,11 @@ else(CLR_CMAKE_HOST_WIN32)
     if(CLR_CMAKE_TARGET_LINUX OR CLR_CMAKE_TARGET_FREEBSD OR CLR_CMAKE_TARGET_NETBSD OR CLR_CMAKE_TARGET_SUNOS)
         # This option is necessary to ensure that the overloaded delete operator defined inside
         # of the utilcode will be used instead of the standard library delete operator.
-        add_link_options("LINKER:-Bsymbolic")
+        if (CMAKE_CXX_COMPILER_ID STREQUAL Clang)
+            add_linker_flag("-Xlinker -Bsymbolic")
```
Comment (janvorli) [inline src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt:37]: -Wl syntax is fine for clang too

File: eng/native/configureplatform.cmake [hunk 5]
```diff
@@ -381,19 +380,9 @@ if(NOT CLR_CMAKE_TARGET_BROWSER)
...
-        else()
-            set(CLR_PIE_LANGUAGE C)
-        endif()
-        check_pie_supported(OUTPUT_VARIABLE PIE_SUPPORT_OUTPUT LANGUAGES ${CLR_PIE_LANGUAGE})
-        if(NOT MSVC AND NOT CMAKE_${CLR_PIE_LANGUAGE}_LINK_PIE_SUPPORTED)
-            message(WARNING "PIE is not supported at link time: ${PIE_SUPPORT_OUTPUT}.\n"
-                      "PIE link options will not be passed to linker.")
-        endif()
+    if(NOT CLR_CMAKE_TARGET_ANDROID AND NOT CLR_CMAKE_TARGET_SUNOS AND NOT MSVC)
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fpie")
```
Comment (janvorli) [inline eng/native/configureplatform.cmake:384]: I think this is not correct. The linker option is just -pie for exes and nothing for shared libs. We also need to add -fPIC and -fPIE compiler options.

File: eng/native/build-commons.sh [hunk 6]
```diff
@@ -48,11 +48,6 @@ check_prereqs()
 {
     echo "Checking prerequisites..."
 
-    if ! cmake --help 2>&1 | grep -q \\-B; then
-        echo "Please install cmake v3.14.5 or newer from https://www.cmake.org/download/."
```
Comment (akoeplinger) [inline eng/native/build-commons.sh:52]: I think it'd still be good to keep this check since it makes the build experience much nicer when you're missing cmake (assuming the old cmake returns some parsable version number from `--help`).

File: eng/native/configureplatform.cmake [hunk 7]
```diff
@@ -381,19 +380,10 @@ if(NOT CLR_CMAKE_TARGET_BROWSER)
...
     # The default linker on Solaris also does not support PIE.
-    if(NOT CLR_CMAKE_TARGET_ANDROID AND NOT CLR_CMAKE_TARGET_SUNOS)
-        # All code we build should be compiled as position independent
-        get_property(languages GLOBAL PROPERTY ENABLED_LANGUAGES)
-        if("CXX" IN_LIST languages)
-            set(CLR_PIE_LANGUAGE CXX)
-        else()
-            set(CLR_PIE_LANGUAGE C)
-        endif()
-        check_pie_supported(OUTPUT_VARIABLE PIE_SUPPORT_OUTPUT LANGUAGES ${CLR_PIE_LANGUAGE})
```
Comment (akoeplinger) [inline eng/native/configureplatform.cmake:392]: there's a comment about `check_pie_supported` at the top that no longer applies

File: src/coreclr/pgosupport.cmake [hunk 8]
```diff
@@ -1,5 +1,18 @@
-include(CheckIPOSupported)
-check_ipo_supported(RESULT HAVE_LTO)
+include(CheckCXXSourceCompiles)
```
Comment (akoeplinger) [inline src/coreclr/pgosupport.cmake:1]: nit: all lines in this file have an extra `tab` character at the end of the line

File: src/coreclr/src/ToolBox/SOS/DacTableGen/CMakeLists.txt [hunk 9]
```diff
@@ -1,3 +1,4 @@
+cmake_minimum_required(VERSION 3.8)
```
Comment (akoeplinger) [inline src/coreclr/src/ToolBox/SOS/DacTableGen/CMakeLists.txt:1]: should this be 3.6.2 as well here?

Comment (akoeplinger) [reply src/coreclr/src/ToolBox/SOS/DacTableGen/CMakeLists.txt:1]: actually I see a few other files that use 3.6.3 instead, we should be consistent.

File: src/libraries/Native/Unix/CMakeLists.txt [hunk 10]
```diff
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.14.5)
+cmake_minimum_required(VERSION 3.6.3)
```
Comment (akoeplinger) [inline src/libraries/Native/Unix/CMakeLists.txt:1]: can you please add a check for cmake > 3.14.5 when targetting iOS/tvOS like we had in an earlier version of the file: https://github.com/dotnet/runtime/blob/a56d6a8034f5cb5ad58f9e29eb61edcc792f4b86/src/libraries/Native/Unix/CMakeLists.txt#L8-L11

File: src/coreclr/src/ToolBox/SOS/DacTableGen/CMakeLists.txt (hunk 9 above)
Comment (jkoritzinsky) [reply src/coreclr/src/ToolBox/SOS/DacTableGen/CMakeLists.txt:1]: This particular project (which is Windows only) needs CMake 3.8's C# support. This check ensures we don't try building it on platforms with 3.6 only.

File: eng/native/build-commons.sh (hunk 6 above)
Comment (jkoritzinsky) [reply eng/native/build-commons.sh:52]: This check was also being hit by the RedHat build, so either we'd need to branch this check for Linux/OSX, or since we aren't using the `-B` and `-S` flags any more, we can just let CMake do the version check when it starts configuring.

**Comment by omajid** [comment]
Thanks for this PR!

I tried building `ad16a9797a2e899caec8a610f667789e0bec6524`  using `./build.sh` on RHEL 8. It doesn't seem to work.

```
  Detected Linux x86_64                                                                                                                                                                                                                       
  -- Performing Test COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG                                                                                                                                                                                 -- Performing Test COMPILER_SUPPORTS_F_STACK_PROTECTOR_STRONG - Success                                                                                                                                                                       -- The ASM compiler identification is Clang                                                                                                                                                                                                 
  -- Found assembler: /usr/bin/clang                                                                                                                                                                                                          
  CMake Error at cli/exe.cmake:7 (cmake_policy):                                                                                                                                                                                                  Policy "CMP0083" is not known to this version of CMake.                                                                                                                                                                                   
  Call Stack (most recent call first):                                                                                                                                                                                                            cli/apphost/static/CMakeLists.txt:41 (include)                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            
  -- Performing Test HAVE_DIRENT_D_TYPE                                                                                                                                                                                                         -- Performing Test HAVE_DIRENT_D_TYPE - Success                                                                                                                                                                                             
  Looking for coreclr_static lib at location: '/home/dotnet/runtime/artifacts/bin/coreclr/Linux.x64.Debug//lib'.                                                                                                                                CMake Error at cli/exe.cmake:7 (cmake_policy):                                                                                                                                                                                              
    Policy "CMP0083" is not known to this version of CMake.                                                                                                                                                                                     Call Stack (most recent call first):                                                                                                                                                                                                        
    cli/apphost/standalone/CMakeLists.txt:39 (include)                                                                                                                                                                                                                                                                                                                                                                                                                                      
                                                                                                                                                                                                                                                CMake Error at cli/exe.cmake:7 (cmake_policy):                                                                                                                                                                                              
    Policy "CMP0083" is not known to this version of CMake.                                                                                                                                                                                   
  Call Stack (most recent call first):                                                                                                                                                                                                        
    cli/dotnet/CMakeLists.txt:16 (include)                                                                                                                                                                                                    
                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                              
  CMake Error at cli/fxr/standalone/CMakeLists.txt:51 (target_link_libraries):                                                                                                                                                                
    Target "libhostfxr_static" of type OBJECT_LIBRARY may not be linked into                                                                                                                                                                  
    another target.  One may link only to INTERFACE, STATIC or SHARED                                                                                                                                                                         
    libraries, or to executables with the ENABLE_EXPORTS property set.                                                                                                                                                                        
                                                                                                                                                                                                                                              
                                                                                                                                                                                                                                              
  CMake Error at cli/test/mockhostfxr/CMakeLists.txt:4 (cmake_minimum_required):                                                                                                                                                              
    CMake 3.14.5 or higher is required.  You are running version 3.11.4                                                                                                                                                                       
                                                                                                                                                                                                                                              
   
  -- Configuring incomplete, errors occurred!
```

Please let me know if there's anything I can do to help.

**Comment by jkoritzinsky** [comment]
I've fixed the issues that I've found relatively simple (removing the unknown policy and removing the extra cmake_minimum_required), but I'm not sure how to fix the `libhostfxr_static` linking issue. If you know how to downgrade that, I'd appreciate the help.

**Comment by omajid** [comment]
> If you know how to downgrade that, I'd appreciate the help.

Not off the top of my head. I will dig around.

**Comment by janvorli** [comment]
> but I'm not sure how to fix the `libhostfxr_static` linking issue

The object libraries cannot be used as regular libraries to link to prior to cmake 3.14. All places need to be changed so that the objects from the object libraries are added to the respective targets like source files. E.g.:
```
add_library(... $<TARGET_OBJECTS:objlib> ...)
```

**Comment by jashook** [comment]
I am having trouble building with 3.6.2. Seems that https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/eventing/CMakeLists.txt#L11 requires cmake 3.8+.

How did you get around this @jkoritzinsky

**Comment by jkoritzinsky** [comment]
I forgot about that. I need to roll that back to use the old find python module.

**Comment by jkoritzinsky** [comment]
@safern can you take a look at the libraries build failures on OSX?

**Comment by jashook** [comment]
```
-- Looking for include file sys/mman.h
-- Looking for include file sys/mman.h - found
-- Performing Test HAVE_PTHREAD_THREADID_NP
-- Performing Test HAVE_PTHREAD_THREADID_NP - Failed
-- Performing Test HAVE_PTHREAD_GETTHREADID_NP
-- Performing Test HAVE_PTHREAD_GETTHREADID_NP - Failed
-- Looking for process_vm_readv
-- Looking for process_vm_readv - found
CMake Error at src/utilcode/CMakeLists.txt:108 (target_link_libraries):
  Object library target "utilcode" may not link to anything.


-- Configuring incomplete, errors occurred!
See also "/runtime/artifacts/obj/coreclr/Linux.x64.Release/CMakeFiles/CMakeOutput.log".
See also "/runtime/artifacts/obj/coreclr/Linux.x64.Release/CMakeFiles/CMakeError.log".
/runtime
```

@janvorli any opinions on how to address the newest object library issue?

**Comment by janvorli** [comment]
> @janvorli any opinions on how to address the newest object library issue?

I've already mentioned that in my comment above. The object libraries cannot be used in target_link_libraries, they need to be put to add_library or add_executable.

**Comment by jkoritzinsky** [comment]
This one is slightly different. In this case we're producing an object library, not trying to consume one. Is it the same resolution?

**Comment by janvorli** [comment]
Yes, I believe it should be done the same way.

**Comment by omajid** [comment]
With one more patch, I can get `./build.sh clr` to work on RHEL 8 with `cmake version 3.11.4`:

```diff
diff --git a/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt b/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
index 8cfd676aecf..2a25b2119b3 100644
--- a/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
+++ b/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
@@ -109,7 +109,7 @@ set(CORECLR_LIBRARIES
     ildbsymlib
     utilcode
     v3binder
-    System.Globalization.Native-Static
+    System.Globalization.Native-static
     interop
 )
 
```

**Comment by omajid** [comment]
With this patch, I can build all of runtime using cmake 3.11.4:

```diff
diff --git a/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt b/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
index 8cfd676aecf..2a25b2119b3 100644
--- a/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
+++ b/src/coreclr/src/dlls/mscoree/coreclr/CMakeLists.txt
@@ -109,7 +109,7 @@ set(CORECLR_LIBRARIES
     ildbsymlib
     utilcode
     v3binder
-    System.Globalization.Native-Static
+    System.Globalization.Native-static
     interop
 )
 
diff --git a/src/installer/corehost/cli/apphost/static/CMakeLists.txt b/src/installer/corehost/cli/apphost/static/CMakeLists.txt
index eb8c4ad6795..47f92e613cf 100644
--- a/src/installer/corehost/cli/apphost/static/CMakeLists.txt
+++ b/src/installer/corehost/cli/apphost/static/CMakeLists.txt
@@ -31,6 +31,7 @@ set(HEADERS
 )
 
 list(APPEND SOURCES $<TARGET_OBJECTS:libhostfxr_static>)
+list(APPEND SOURCES $<TARGET_OBJECTS:libhostpolicy_static>)
 
 if(CLR_CMAKE_TARGET_WIN32)
     list(APPEND SOURCES
@@ -140,7 +141,6 @@ endif(CLR_CMAKE_HOST_UNIX)
 set_property(TARGET singlefilehost PROPERTY ENABLE_EXPORTS 1)
 
 target_link_libraries(singlefilehost
-    libhostpolicy_static
     libhostcommon
     ${CORECLR_LIBRARIES}
 )
diff --git a/src/installer/corehost/cli/fxr/standalone/CMakeLists.txt b/src/installer/corehost/cli/fxr/standalone/CMakeLists.txt
index d6c147f8f58..3e1a6d04d37 100644
--- a/src/installer/corehost/cli/fxr/standalone/CMakeLists.txt
+++ b/src/installer/corehost/cli/fxr/standalone/CMakeLists.txt
@@ -48,4 +48,4 @@ if(CLR_CMAKE_HOST_UNIX)
 endif(CLR_CMAKE_HOST_UNIX)
 
 install_with_stripped_symbols(hostfxr TARGETS corehost)
-target_link_libraries(hostfxr libhostcommon libhostfxr_static)
+target_link_libraries(hostfxr libhostcommon $<TARGET_OBJECTS:libhostfxr_static>)
```

@jkoritzinsky Any chance you could try this out?

**Comment by jkoritzinsky** [comment]
I've applied @omajid's patch and removed the `install` commands for object libraries based on offline conversations. I believe we are now able to configure CMake 3.6.2. I've seen some local failures in the PAL while building with the image @jashook provided. @janvorli since you know the area best, can you take a look?

**Comment by janvorli** [comment]
> @janvorli since you know the area best, can you take a look?

Sure, let me take a look

**Comment by janvorli** [comment]
Hmm, I have synced this branch and the file was generated properly, but it failed with this error. I've verified that the respective binary folder with the header is passed to the compiler using the -I option.
Re-running the build (without cleaning) actually makes the build complete, so I am convinced the problem is caused by incorrect dependency tracking for the header in the cmake scripts.

**Comment by janvorli** [comment]
Maybe the problem is that EventingHeaders in vm/eventing/CMakeLists.txt don't contain the clrproviders.h

**Comment by janvorli** [comment]
Ok, so the problem is elsewhere. We are not adding eventing_headers as a dependency to clr_wks_obj and cee_wks_mergeable_obj in src/coreclr/src/vm/wks/CMakeLists.txt. After adding them to the add_dependencies commands, it builds ok.

**Comment by jkoritzinsky** [comment]
@janvorli if you already have a fix locally, can you push it out to this branch? You should have permissions to do so.

**Comment by janvorli** [comment]
@jkoritzinsky - done

**Comment by jkoritzinsky** [comment]
I cannot seem to figure out why CMake isn't recognizing the custom-target-compiled assembly objects as generated. It's causing the Windows build failures.

**Comment by jkoritzinsky** [comment]
@janvorli any chance you can help out? I'm not quite sure how this broke today and I'm out of ideas on how to fix it.

**Comment by janvorli** [comment]
> I cannot seem to figure out why CMake isn't recognizing the custom-target-compiled assembly objects as generated.

Looking at the error:
```
  Cannot find source file:

    F:/workspace/_work/1/s/artifacts/obj/coreclr/Windows_NT.arm.Checked/src/vm/wks/cee_wks_core.dir/$(Configuration)/asmconstants.obj
```

The asmconstants should not produce an object file, as it is just a .h file transformed into .inc file that is included by all the .asm files. So it looks like we are somehow accidentally trying to compile it standalone somehow.

I can take a look tomorrow if you won't figure it out before that.

**Comment by janvorli** [comment]
I've pushed a fix. The problem was that the VM_HEADERS_WKS_ARCH_ASM that contained the asmconstants.h was being added to VM_SOURCES_WKS_ARCH_ASM, which is a set of files that are to be compiled with assembler. It needed to be added to VM_SOURCES_WKS instead (which is done just to let the header appear in the visual studio solution explorer, cmake is clever enough to not to pass .h files to the C++ compiler).

**Comment by jkoritzinsky** [comment]
@omajid can you do a test build against the RHEL 7 image you'll be building against for the distro build?

**Comment by omajid** [comment]
> @omajid can you do a test build against the RHEL 7 image you'll be building against for the distro build?

Of course. I have tested on RHEL 8 using this PR (it works!). I am currently testing on RHEL 8 end-to-end now (using source-build Preview 4 + a manual backport of this patch). If/when that works, RHEL 7 is next on my list.

**Comment by omajid** [comment]
> @omajid can you do a test build against the RHEL 7 image you'll be building against for the distro build?

A `./build.sh clr` works with commit `265f684da4d370909c70399fd9be4bd089c607b2` after this trivial change:

```diff
diff --git a/src/installer/corehost/CMakeLists.txt b/src/installer/corehost/CMakeLists.txt
index 178d8ab..46d1c5e 100644
--- a/src/installer/corehost/CMakeLists.txt
+++ b/src/installer/corehost/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.6.3)
+cmake_minimum_required(VERSION 3.6.2)
 
 project(corehost)
 
diff --git a/src/libraries/Native/Unix/CMakeLists.txt b/src/libraries/Native/Unix/CMakeLists.txt
index dee9e51..3168c7e 100644
--- a/src/libraries/Native/Unix/CMakeLists.txt
+++ b/src/libraries/Native/Unix/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.6.3)
+cmake_minimum_required(VERSION 3.6.2)
 if(CLR_CMAKE_TARGET_IOS OR CLR_CMAKE_TARGET_TVOS) 
      # CMake 3.14.5 contains bug fixes for iOS 
      cmake_minimum_required(VERSION 3.14.5) 
diff --git a/src/tests/CMakeLists.txt b/src/tests/CMakeLists.txt
index c75e5db..77fcc16 100644
--- a/src/tests/CMakeLists.txt
+++ b/src/tests/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required(VERSION 3.6.3)
+cmake_minimum_required(VERSION 3.6.2)
 
 cmake_policy(SET CMP0042 NEW)
 project(Tests)
```

**Comment by jkoritzinsky** [comment]
I've applied your patch. Marking this PR as ready for review/non-work in progress

**Review by akoeplinger — APPROVED** [review]
The non-coreclr specific changes look good to me (can't vouch for the coreclr changes).

File: src/coreclr/src/classlibnative/bcltype/CMakeLists.txt [hunk 11]
```diff
@@ -10,9 +10,11 @@ set(BCLTYPE_SOURCES
...
 
-add_library_clr(bcltype
+add_library_clr(bcltype_obj
     OBJECT
     ${BCLTYPE_SOURCES}
 )
 
-add_dependencies(bcltype eventing_headers)
+add_dependencies(bcltype_obj eventing_headers)
+add_library(bcltype INTERFACE)
```
Comment (janvorli) [inline src/coreclr/src/classlibnative/bcltype/CMakeLists.txt:19]: An idea for a future cleanup: This add_library / target_sources pair pattern is used at 24 places. It may be nice to clean this up in a follow up change by creating a cmake function for this pattern. Or even encapsulate the add_library_clr in that function, getting something along these lines:
```cmake
add_intermediate_library_clr(bcltype ${BCLTYPE_SOURCES})
add_intermediate_library_dependencies(bcltype eventing_headers)
```

File: src/installer/corehost/cli/apphost/static/CMakeLists.txt [hunk 12]
```diff
@@ -30,6 +30,8 @@ set(HEADERS
     ../../../hostfxr_resolver.h
 )
 
+list(APPEND SOURCES $<TARGET_OBJECTS:libhostfxr_static> $<TARGET_OBJECTS:libhostpolicy_static>)
```
Comment (janvorli) [inline src/installer/corehost/cli/apphost/static/CMakeLists.txt:33]: I wonder why was this change needed.  Linking to the static libraries should work fine. The only case when we need to use object libraries is when a static library is built as a combination of other libraries, but it is not the case here.

File: src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt [hunk 13]
```diff
@@ -80,14 +80,15 @@ if (GEN_SHARED_LIB)
     install_with_stripped_symbols (System.Globalization.Native PROGRAMS .)
 endif()
 
-add_library(System.Globalization.Native-Static
+add_library(System.Globalization.Native-obj
```
Comment (janvorli) [inline src/libraries/Native/Unix/System.Globalization.Native/CMakeLists.txt:83]: A nit - for other libraries, we use `_obj` suffix to indicate object libraries, so it would be nice to stay consistent here.

File: src/installer/corehost/cli/apphost/static/CMakeLists.txt (hunk 12 above)
Comment (jkoritzinsky) [reply src/installer/corehost/cli/apphost/static/CMakeLists.txt:33]: The `libhostfxr_static` and `libhostpolicy_static` targets are object libraries. The corehost CMake infra could really use some updates and cleanup to be more straightforward. I'd like to do that in a future PR.

Comment (janvorli) [reply src/installer/corehost/cli/apphost/static/CMakeLists.txt:33]: Hmm, interesting. I can see that they were changed to object ones few months ago and I am not sure why.

I agree that the corehost cmakefiles are very difficult to mentally parse. They looks as if msbuild style of build programming was used here. Instead of using functions for doing common stuff, ambient cmake variables are being set and then common helper cmake files included. I'd love to see it transformed to a natural cmake way at some point.

File: src/coreclr/src/classlibnative/bcltype/CMakeLists.txt (hunk 11 above)
Comment (jkoritzinsky) [reply src/coreclr/src/classlibnative/bcltype/CMakeLists.txt:19]: Yes, we should clean this up in the future by making some additional functions.

**Comment by jeffschwMSFT** [comment]
cc @mmitche

**Comment by jkoritzinsky** [comment]
@janvorli any more comments?

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Comment by janvorli** [comment]
@jkoritzinsky no, I have thought I've already approved it, but I can see I haven't. So I've just done that.

**Comment by jashook** [comment]
Thank you @jkoritzinsky for driving this effort!

**Comment by jashook** [comment]
@omajid which version of clang are we using on redhat 7 for source build? We noticed that by default centos 7 installs 3.4 which we do not support building with.

/cc @janvorli @jkoritzinsky

**Comment by omajid** [comment]
>  which version of clang are we using on redhat 7 for source build?

We have generally used `llvm-toolset-7.0-cmake` ([CentOS version here](https://cbs.centos.org/koji/packageinfo?packageID=7417)), which is 3.6.2.

**Comment by jashook** [comment]
I have a clean build using clang/llvm 7 with the changes on centos 7

**Comment by mmitche** [comment]
@jeffschwMSFT Was this supposed to be ported to p8?

**Comment by jeffschwMSFT** [comment]
@mmitche yes, @jashook is working on the port.

**Comment by omajid** [comment]
I did a port to preview 4, so I could build against older cmake versions using source-build. Source-build is currently at 5.0 Preview 4. Port is available here: https://github.com/omajid/runtime/tree/5.0-preview4-cmake-backport
